### ISSUE TITLE
[Bug]: Electron ignores XCompose settings when running natively on Wayland

### ISSUE THREAD

**Issue by gnull** [electron/electron#29345]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Electron Version

12.0.7

### What operating system are you using?

Other Linux

### Operating System Version

Arch Linux

### What arch are you using?

x64

### Expected Behavior

1. Start a [Sway](https://swaywm.org/) session.

1. Run `swaymsg 'input * xkb_options "compose:menu"'` to enable XCompose on the [`Menu`](https://en.wikipedia.org/wiki/Menu_key) key. (See `man 7 xkeyboard-config | grep 'compose:'` for keys other than `Menu` that you can use.)

1. Put

   ```bash
   --enable-features=UseOzonePlatform
   --ozone-platform=wayland
   ```

   into your `~/.config/electron-flags.conf`.

1. Run the [Hash](https://www.electronjs.org/blog/simple-samples#hash) sample Electron app. I.e. do

   ```bash
   git clone https://github.com/electron/simple-samples .
   cd hash
   npm start
   ```

1. Focus on the first input field in the Hash app window.

1. Input symbols using XCompose:

   1. Press the following in sequence: `Menu` `-` `-` `-`. It must produce the em-dash symbol in the input field, “`—`”.

   1. Press the following in sequence: `Menu` `=` `>`. It must produce the double arrow symbol, “`⇒`”.

### Actual Behavior

The last step does not produce any of the expected symbols in the input field. Instead of em-dash it produces three sort minus signs “`---`”, and instead of double arrow it produces “`=>`”.

### Additional Information

Everything works as expected and XCompose produces “`—`” and “`⇒`” if I omit the step 3, i.e. run Electron without the `--enable-features=UseOzonePlatform --ozone-platform=wayland` options.

**Comment by gnull** [comment]
I also noticed that pasting text with middle mouse button doesn't work. To reproduce, do the steps 1-6 as above. Instead of step 7 do, run `echo kek | wl-copy --primary` in terminal and then click with the middle mouse button on the input field in the Hash app. The expected behaviour is “kek” appearing in the input field, but nothing happens.

**Comment by gnull** [comment]
I can't reproduce this issue anymore in Electron version `v13.1.4`. Both XCompose and middle click paste work as expected for me with Sway `1.6`.

**Comment by gnull** [comment]
Oops, my previous comment is wrong. The middle click paste works, but XCompose combinations don't.

**Comment by gnull** [comment]
I checked this again on version `v13.1.9`. The middle click paste works, but the XCompose combinations still don't.

**Comment by gnull** [comment]
Good news! The issue appears to be fixed in Electron `v13.4.0`, all the XCompose sequences work for me.

But in Electron `v14.0.1` the issue is still present, XCompose sequences do not work.

The middle-click paste works in both versions.

**Comment by MaxIhlenfeldt** [comment]
A fix for this has recently been merged upstream (https://crrev.com/c/3207928) and will be included in Chromium 98 (release scheduled for Feb 1, 2022).

**Comment by RaisinTen** [comment]
@gnull @MaxIhlenfeldt could y'all test it on Electron v17, v18 and v19? All of these are on versions of Chromium that are >= M98, so my hunch is that this issue has been fixed already but it would be nice to have a confirmation!

**Comment by Myrdden** [comment]
@RaisinTen Hello! Just upgraded an app to v18, haven't gotten it working with v19 yet. I can confirm that while the compose key itself is working fine on v18, nothing specified in my ~/.XCompose file works, only the default bindings.

**Comment by RaisinTen** [comment]
Hi @Myrdden, thank you for the update! There seems to be a related Chromium bug report - https://bugs.chromium.org/p/chromium/issues/detail?id=506838 that talks about how the `~/.XCompose` file is ignored. I guess we need to wait for someone to send a Chromium CL similar to https://codereview.chromium.org/243143002 to add support for reading settings from the `~/.XCompose` file, just like it currently works on X11.

**Comment by RalfJung** [comment]
I am still affected by this in vscodium. (Not sure how to figure out which electron version that is.)

**Comment by noonsleeper** [comment]
> I am still affected by this in vscodium. (Not sure how to figure out which electron version that is.)

Hey @RalfJung

Open vscodium go to help menu > about

**Comment by RalfJung** [comment]
Wow that bot is aggressive. When the devs don't have the time to even notice an issue we have to post every 3 months that the issue did not miraculously fix itself? Sorry, that's ridicolous.

**Comment by yue-dongchen** [comment]
@RalfJung This seems to be a problem with Chromium where they [switched to internal keycode-to-keysym mappings disregarding modifier keys to compensate for non-UTF locale settings](https://bugs.chromium.org/p/chromium/issues/detail?id=395019). Of course then it caused problems as [this Chromium issue from 2014 raises (xmodmap)](https://bugs.chromium.org/p/chromium/issues/detail?id=394552).

Indeed the compose key is probably [fixed](https://github.com/electron/electron/issues/29345#issuecomment-963027377), but the modifier state is still ignored.

As a result this also affects other Chromium-based apps like I've noticed with Onlyoffice.

**Comment by RalfJung** [comment]
For me the compose key is not fixed. I have a lot of custom compose sequences in `~/.XCompose`, and Electron applications ignore them when running natively on Wayland. I have to run them all in XWayland so they function properly.

**Comment by MaxIhlenfeldt** [comment]
For the record, the upstream issue is https://crbug.com/1455511.

**Comment by RalfJung** [comment]
I can't reply there (no Google account), but this part is not accurate:

> as .XCompose is specific to X11, we (Chromium) probably won't support it on Wayland

.XCompose works on Wayland as well. Though maybe this is the answer:

> is most or all of these work because they support "ibus", which does the actual reading of the .XCompose file

If they close the XCompose issue, I hope they open some other issue to track proper input method support. The current support is just not sufficient for productive use for some people.

**Comment by MaxIhlenfeldt** [comment]
> If they close the XCompose issue, I hope they open some other issue to track proper input method support. The current support is just not sufficient for productive use for some people.

Chromium's IME support is currently blocked on Wayland protocol issues. https://crbug.com/1039161 has more details, if you're interested.

**Comment by cosme-benito** [comment]
Is there a workaround, at least?

**Comment by MaxIhlenfeldt** [comment]
Not sure how applicable this is to Electron, but you might get Chromium to support IME and use ibus by starting it with `--gtk-version=4`.

**Comment by cosme-benito** [comment]
Unfortunately for Electron apps such as Discord this doesn't work. 

I rely on xcompose to fix my dead keys, it's incredibly frustrating to type in english without them (example: when trying to type `I'm` the result is actually `Iḿ` .

**Comment by ViciousBadger** [comment]
Still an issue in Electron v27 unfortunately

**Comment by ariasuni** [comment]
Still an issue on Electron 36. Somehow Discord started using Wayland by default so I’m stuck with being unable to use correctly my keyboard layout… Fixed by using e.g. `ELECTRON_OZONE_PLATFORM_HINT=x11 discord`.

Edit: I tried both Discord with whatever Electron version they ship, and a custom version using Electron 36.

Edit: Still an issue with Electron 39.
